FIX: Tensorboard callback only supports logging Embeddings layer weights

dschwertfeger opened the pull request:
Currently the Tensorboard embeddings visualizer only visualizes layer weights, rather than the layer outputs. This issue has already been raised twice:

- https://github.com/fchollet/keras/issues/6674
- https://github.com/fchollet/keras/issues/6287

This PR fixes this by actually _embedding_ the provided `embeddings_data` using the layer weights of the given `embeddings_layer_names`.

What has changed?

`keras.callbacks.TensorBoard()` accepts an additional parameter: `embeddings_data`. In practice, this is often the same as the validation data (`x_val`) but it could be any arbitrary data. The number of samples in `embeddings_data` is needed to reshape the embedding's layer's output accordingly.

Before:
<img width="1440" alt="before" src="https://user-images.githubusercontent.com/1120378/30173742-0d1ee5f0-93f9-11e7-816e-023af2d5ba8b.png">

After:
<img width="1440" alt="after" src="https://user-images.githubusercontent.com/1120378/30173756-14fe614c-93f9-11e7-8949-8d9cb458e139.png">

fchollet commented:
Thanks for the PR.

Could someone help review this PR, and determine if this change is needed (or if this is the best way to handle it)?

ahundt commented:
could you include a before/after screenshot since this changes what tensorboard does?

dschwertfeger commented:
Exactly, the `validation_data_size` refers to the number of samples in the validation data. This is needed to create the `tf.Variable` to hold the `embedding`s to ultimately visualize each validation-data sample at the output of the embedding layer. Before, each weight at the "embedding" layer was visualized (see "before"-screenshot above), which doesn't make sense.

Haven't come across `steps` yet, so I'm not sure about that case. I can say, though, we definitely need to know the number of samples in the validation data if we want to visualize their embedding at a given layer.

Is there any way to find out the number of samples in the `validation_data` when using steps?

chleibig commented:
Thanks a lot @dschwertfeger for this PR. I can confirm the need for this and with minor changes (see below) it works for my purposes. Things I noticed:
* In principle, the `validation_data_size` could be inferred from the callback's `validation_data`, however   the latter is only set after `set_model` needs it. This and how to deal with the `steps` case, seems related, but unfortunately I didn't have the time to dig deeper so far.
* within `on_epoch_end`:
      * The learning phase is not set, but needed to work with e.g. dropout layers.
      *  The embeddings are computed for all the validation data at once instead of iterating over batches.

chleibig commented:
What I could also imagine to be useful is to have a kwarg called `embedding_data` instead of `validation_data_size`. This would be in line with `embedding_metadata` and allow to feed any data for embedding, be it the validation data, a subset thereof or something else. Furthermore, I currently don't see an issue with `steps`.

fchollet commented:
Thanks for the feedback @chleibig ! Additionally, is there a way to remove the argument `validation_data_size`? It seems redundant, and I don't think it is a good user experience to have to specify it.

dschwertfeger commented:
Hi @fchollet , 

I agree it's redundant but we need to know the number of samples in the data to be embedded when we instantiate the `TensorBoard` callback. (Well, in `set_model` to be precise.)

If we explicitly pass the `embedding_data` to the callback as @chleibig suggested, we can get rid of the `validation_data_size`. 

Would that work for you?

dschwertfeger commented:
Thanks for your comments @chleibig . I'm not sure about this statement

> within `on_epoch_end` the learning phase is not set, but needed to work with e.g. dropout layers

Could you please elaborate?

Also, I think it's totally fine to compute the embeddings for all data at once. What's your concern with this?

chleibig commented:
Hi @dschwertfeger,
sure.
* Regarding the learning phase: keras uses a boolean flag (you can get the `placeholder` via `K.learning_phase()`) that indicates whether a model is used in training (e.g. dropout turned on, learning phase = 1) or testing/evaluation (e.g. dropout turned off, learning_phase = 0) phase. I don't know where this is documented, but if you go through the code base, you will find further ways how to use it, e.g. in analogy to your contribution in `on_epoch_end` within the `if self.validation_data and self.histogram_freq` clause. For your case you could as well do something like:
    ```python
    if self.model.uses_learning_phase:
        feed_dict[K.learning_phase()] = 0
   ```
* Regarding the iteration over batches: Imagine the following scenario: you want to embed a few thousand data points let's say in the last hidden layer of a deep vision model. While the `embedding` of shape (n_samples, n_hidden) may still fit in memory, intermediate representations of higher dimensionality may not. Hence performing the forward passes in blocks of `self.batch_size` might be necessary. Underneath the above mentioned clause `if self.validation_data and self.histogram_freq` the data is processed as well in batches.

I hope this helps.

mhajiaghayi commented:
I'm trying to visualize the embedding layer output for the imdb sentiments in Keras. But I don't get the expected result. Specifically, I don't see the expected relation among words, e.g. word kingdom is not near king or queen. can this issue be related to what you have mentioned above? I get a good accuracy for the model. so I suppose the embedding should be fine too. I have attached the code here. 

```python
from __future__ import print_function
from keras.preprocessing import sequence
from keras.models import Sequential
from keras.layers import Dense, Embedding
from keras.layers import LSTM
from keras.datasets import imdb
from keras.callbacks import TensorBoard
from keras.callbacks import ModelCheckpoint
import pdb,shutil
import tensorflow as tf 
max_features = 1000
maxlen = 80  # cut texts after this number of words (among top max_features most common words)
batch_size = 32


def saveActionMetaData(_map, fileName,max_features):
    row = 0
    with open(fileName, 'w') as f:
        f.write("Word\tFrequency\n")
        for action, _id in _map.items():
            try:
                if _id <= max_features:
                    if row < max_features:
                        f.write(action + "\t" +str(_id) + "\n")
                    else: 
                        f.write(action + "\t" +str(_id))
                    row = row + 1 
            except:
                print("something is wrong")
                print(action)
                continue 
    return

with tf.device('cpu:0'): 
    ROOT_DIR = 'tfboard'
    shutil.rmtree(ROOT_DIR, ignore_errors=True)
    print('Loading data...')
    (x_train, y_train), (x_test, y_test) = imdb.load_data(num_words=max_features)
    _map = imdb.get_word_index()
    saveActionMetaData(_map, 'metadata.tsv',max_features)
    print(len(x_train), 'train sequences')
    print(len(x_test), 'test sequences')
    print('Pad sequences (samples x time)')
    x_train = sequence.pad_sequences(x_train, maxlen=maxlen)
    x_test = sequence.pad_sequences(x_test, maxlen=maxlen)
    print('x_train shape:', x_train.shape)
    print('x_test shape:', x_test.shape)

    print('Build model...')
    model = Sequential()
    model.add(Embedding(max_features, 40, name = 'embedding'))
    model.add(LSTM(40, dropout=0.2, recurrent_dropout=0.2))
    model.add(Dense(1, activation='sigmoid'))

    # try using different optimizers and different optimizer configs
    model.compile(loss='binary_crossentropy',
                  optimizer='adam',
                  metrics=['accuracy'])

    print('Train...')


    embeddingsMetadata = {'embedding': 'metadata.tsv'}
    tbCallback = TensorBoard( log_dir= ROOT_DIR,
                            histogram_freq= 10, 
                            write_graph = False,
                            embeddings_freq=10, 
                            batch_size=32,
                            embeddings_layer_names = ['embedding'],
                            embeddings_metadata= embeddingsMetadata
                             )
    model.fit(x_train, y_train,
              batch_size=batch_size,
              epochs=40,
              validation_data=(x_test, y_test),
              callbacks = [tbCallback],
              verbose = 1 )
    score, acc = model.evaluate(x_test, y_test,
                                batch_size=batch_size)
    print('Test score:', score)
    print('Test accuracy:', acc)
```

mhajiaghayi commented:
I'm trying to visualize the embedding layer output for the imdb sentiments in Keras. But I don't get the expected result. Specifically, I don't see the expected relation among words, e.g. word kingdom is not near king or queen. can this issue be related to what you have mentioned above? I get a good accuracy for the model. so I suppose the embedding should be fine too. I have attached the code here. 

```python
from __future__ import print_function
from keras.preprocessing import sequence
from keras.models import Sequential
from keras.layers import Dense, Embedding
from keras.layers import LSTM
from keras.datasets import imdb
from keras.callbacks import TensorBoard
from keras.callbacks import ModelCheckpoint
import pdb,shutil
import tensorflow as tf 
max_features = 1000
maxlen = 80  # cut texts after this number of words (among top max_features most common words)
batch_size = 32


def saveActionMetaData(_map, fileName,max_features):
    row = 0
    with open(fileName, 'w') as f:
        f.write("Word\tFrequency\n")
        for action, _id in _map.items():
            try:
                if _id <= max_features:
                    if row < max_features:
                        f.write(action + "\t" +str(_id) + "\n")
                    else: 
                        f.write(action + "\t" +str(_id))
                    row = row + 1 
            except:
                print("something is wrong")
                print(action)
                continue 
    return

with tf.device('cpu:0'): 
    ROOT_DIR = 'tfboard'
    shutil.rmtree(ROOT_DIR, ignore_errors=True)
    print('Loading data...')
    (x_train, y_train), (x_test, y_test) = imdb.load_data(num_words=max_features)
    _map = imdb.get_word_index()
    saveActionMetaData(_map, 'metadata.tsv',max_features)
    print(len(x_train), 'train sequences')
    print(len(x_test), 'test sequences')
    print('Pad sequences (samples x time)')
    x_train = sequence.pad_sequences(x_train, maxlen=maxlen)
    x_test = sequence.pad_sequences(x_test, maxlen=maxlen)
    print('x_train shape:', x_train.shape)
    print('x_test shape:', x_test.shape)

    print('Build model...')
    model = Sequential()
    model.add(Embedding(max_features, 40, name = 'embedding'))
    model.add(LSTM(40, dropout=0.2, recurrent_dropout=0.2))
    model.add(Dense(1, activation='sigmoid'))

    # try using different optimizers and different optimizer configs
    model.compile(loss='binary_crossentropy',
                  optimizer='adam',
                  metrics=['accuracy'])

    print('Train...')


    embeddingsMetadata = {'embedding': 'metadata.tsv'}
    tbCallback = TensorBoard( log_dir= ROOT_DIR,
                            histogram_freq= 10, 
                            write_graph = False,
                            embeddings_freq=10, 
                            batch_size=32,
                            embeddings_layer_names = ['embedding'],
                            embeddings_metadata= embeddingsMetadata
                             )
    model.fit(x_train, y_train,
              batch_size=batch_size,
              epochs=40,
              validation_data=(x_test, y_test),
              callbacks = [tbCallback],
              verbose = 1 )
    score, acc = model.evaluate(x_test, y_test,
                                batch_size=batch_size)
    print('Test score:', score)
    print('Test accuracy:', acc)
```

dschwertfeger commented:
Thanks for the details @chleibig and sorry for the delay. I've been caught up in another project but I'll address your feedback in the coming days.

@mhajiaghayi ,  I can't say for sure but it's very likely related to this issue. If the number of visualized points does not correspond to the number of words you're trying to visualize (but rather to the number of weights in the embedding layer) it's almost certainly it. Until this is fixed here you could use my branch and see if you get the expected results (it's based on v2.7)

fchollet commented on keras/callbacks.py:
```diff
@@ -609,14 +611,16 @@ class TensorBoard(Callback):
             [details](https://www.tensorflow.org/how_tos/embedding_viz/#metadata_optional)
             about metadata files format. In case if the same metadata file is
             used for all embedding layers, string can be passed.
+        embeddings_data: embedding data as Numpy array.
```
Why is this argument necessary? The docstring doesn't describe what it is for or how to use it

Looking at the code, it seems that the data is used as model input to compute embeddings. Why not use the validation data?

On the technical side, this design also makes the assumption that the model has a single input, which maybe not be the case.

dschwertfeger replied:
Thanks for your feedback.

We need to define the `tf.Variable`s to hold the embeddings for every layer in `embeddings_layer_names` inside the `TensorBoard` callback's `set_model` method. The shape of these `embedding_var`s depends on 

1. the number of samples we want to embed (!) and
2. the embedding layer's output dimension

We don't have access to the `validation_data` in the `TensorBoard` callback's `set_model` method, though.

My first approach, up until commit 47f33a6bc411dd956dbfe8975df644d60e0d9666, was to pass an additional argument (`validation_data_size`) and use the `validation_data` to compute the embeddings. You asked to remove it (see above) because it's redundant and wouldn't be a good user experience to specify it. 

I agree.

@chleibig suggested to directly pass the data to be embedded as `embeddings_data` which would allow to embed any arbitrary data, be it

- the training data,
- the validation data,
- a subset thereof or
- something else entirely.

Now, passing `embeddings_data` is also kind of redundant because, in practice, it will probably just be the `validation_data`. I don't know which approach is better or worse but we need to know the number of samples to be embedded when we create the embedding variables in `set_model`. 

I'm happy to add more info about `embeddings_data` to the docstring if this is the way we go.

What do you think?

dschwertfeger replied:
Also, good catch about multiple inputs. I haven't really thought this through but it seems like kind of a special case right now..

**Do you have any specific example use cases mind?**

I could think of these scenarios:

- A siamese network (or similar)
  - multiple inputs of the same kind 
  - sharing the same network
- Shared embeddings
  - for example on images and words
  - multi-label classification using [label embedding](https://arxiv.org/pdf/1312.5650.pdf)
  - ... which I think requires a bit more work than simply taking the output of some arbitrary layer as the "embedding"
- Different inputs types
  - like the Twitter example of the Functional API Guide
    - sequence of words + time of the day
    - ... not sure what to actually visualize in this case
  - or the visual question answering model
- [...]

I think this requires some more thought.

What do you think about tackling this in a separate Issue/PR?

chleibig replied:
Hi together and sorry @dschwertfeger that my suggestion made you having to argue about it;)

I completely agree that the least "redundant" version should be favourable.

For completeness, I list here another reason why I had suggested `embeddings_data`: in terms of style it would be in line with the keyword `embeddings_metadata`,  which provides the labels or other information such as thumbnail images. Furthermore, in case of large (good for evaluation purposes) validation data sets, it may be impractical - due to the memory consumption - to embed all of it. Being able to pass only a subset of the validation data via `embeddings_data` may hence be more desirable.

In case of multiple inputs, `embeddings_metadata` should pass all of them I guess, such as documented e.g. [here](https://keras.io/getting-started/functional-api-guide/).

chleibig commented on keras/callbacks.py:
```diff
@@ -779,11 +796,35 @@ def on_epoch_end(self, epoch, logs=None):
+                        feed_dict[K.learning_phase()] = False
+                        # do not slice the learning phase
+                        batch_val = [x[i:i + step] for x in embeddings_data[:-1]]
+                        batch_val.append(embeddings_data[-1])
```
I don't think `embeddings_data` should have to be passed with the learning phase appended to the end.

chleibig commented on keras/callbacks.py:
```diff
@@ -779,11 +796,35 @@ def on_epoch_end(self, epoch, logs=None):
+                        batch_val = [x[i:i + step] for x in embeddings_data[:-1]]
+                        batch_val.append(embeddings_data[-1])
+                    else:
+                        batch_val = [x[i:i + step] for x in embeddings_data]
```
`batch_val` is not used afterwards. I would have expected it to be passed to the `feed_dict`.

dschwertfeger replied:
Oops. Of course. Thanks for pointing this out. That explains why tests were still green after this change. 

This actually exposes a problem: The `tf.Variable` holding the `embedding` is instantiated with `shape = (n_samples, embedding_dim)`. When I pass the `feed_dict` with only `self.batch_size` samples, this blows up because dimensions (number of samples) are not matched. ☹️

I could not, yet, find a way to save the `tf.Variable` holding the `embedding` in batches.

Theoretically, we could just have a `tf.Variable` for every batch, compute the embedding for `self.batch_size` samples and concatenate them into the final embedding with `n_samples`.

There just doesn't seem to be a way using `tf.train.Saver`.

Any ideas?

ahundt replied:
This probably won't help with your actual problem, but the following PR was very helpful when I needed to fix feed dict issues: https://github.com/fchollet/keras/pull/7064

chleibig replied:
You should assign the embedding_input, which is a batch instead of all data to the appropriate slice of the embeddings variables. I.e. instead of ```tf.assign(embedding, embedding_input)```, you could write ```tf.assign(embeddings[batch_id:batch_id + step], embedding_input)```, with ```batch_id = tf.placeholder(tf.int32)```, that you feed the `i` to when actually iterating over batches. I've written a test with a minimal stand-alone example. Let me know if you want me to paste it here for clarification.

chleibig replied:
Similarly, I could imagine that it might be possible to provide a slice to tf.train.Saver instead of the entire embedding variable.

dschwertfeger replied:
@chleibig, I'd appreciate if you could post your example (if you still have this lying around.)

chleibig replied:
Unfortunately, I don't have this lying around anymore.

ahundt commented:
I finally read through this in detail, and it sounds fantastic! This is exactly one of the things I could benefit from right now, but I do agree a detailed example would be advisable. 

1. Any chance this can take some next steps towards merging?
2. Might it be able to extract some data directly from the internal model by default?

Potentially conflicting (but also very useful) PR of note: https://github.com/keras-team/keras/pull/7617

dschwertfeger commented:
@ahundt, I'd like to see this merged, too. From what I can remember, these are the open issues raised on this PR.

- [ ] Fix data processing in batches 
  - I could make some time to look into this again.
- [ ] Handle multiple inputs
  - Still not sure what exactly the embedding is supposed to visualize if multiple inputs are given.
  - I'd vote for tackling this in a separate PR because it might need more discussion/clarification.

Not quite sure what you mean by "a detailed example would be advisable" and your second point. Can you elaborate?

ahundt commented:
> Fix data processing in batches I could make some time to look into this again.

#7617 does some scalar batch updates, don't know if it is the same kind.

> Still not sure what exactly the embedding is supposed to visualize if multiple inputs are given. I'd vote for tackling this in a separate PR because it might need more discussion/clarification.

This [blog post](https://www.google.com/search?q=embeddings+tensorboard&client=firefox-b-1-ab&ei=xv93WtrzL4ud5wKA7J-QCA&start=10&sa=N&biw=1272&bih=886) seems like it might provide example code and a numpy [save_embeddings()](https://github.com/nlml/np-to-tf-embeddings-visualiser/blob/master/save_embeddings.py) which can work for multiple inputs.


> Not quite sure what you mean by "a detailed example would be advisable" and your second point. Can you elaborate?

Sure, I meant to create a file like `keras/examples/cifar10_embeddings.py` which demonstrates how to use the API.

fchollet commented:
@ahundt @chleibig I will wait for your approval to merge this PR. Thanks for the reviews.

chleibig commented:
Will have a look at the weekend.

chleibig commented:
@dschwertfeger: as the data processing in batches is still open, I have an idea that could maybe work and might even simplify the code and interface: isn't it possible to extract the embeddings (i.e. the hidden layer activations) and create an `embeddings` tensor from the hidden layer activations stored in a numpy array or alike in the very end only?

dschwertfeger commented:
Depends what you mean by "the very end". We can compute the embeddings from a fully trained model. Is that what you mean?

All we need to create the `embeddings` tensor is:

- data to embed and
- the layer name (or names) at which to embed the data.

Curious to hear your idea.

chleibig commented:
For clarification I had the following in mind: underneath `set_model` you currently create `embedding_vars` and `embeddings`, with the latter containing the assign statements that would have to be changed such that you only assign the batch of hidden layer activations to the respective tf.slice into `embedding_vars`. Alternatively, you could get rid of `embeddings` and use the batches (in numpy array format) returned from a `sess.run(...)` in `on_epoch_end` to build up the overall (n_samples, n_hidden) array for each embedding layer and then assign to the `embedding_vars` in one shot with e.g. `K.batch_set_value`. Either of the above two scenarios you have to decide for to get the batch issue working.

The total number of samples is however in both scenarios needed to construct the `embedding_vars`, because this in turn has to be passed to the `tf.train.Saver...` and to build up the `ProjectorConfig`. The first callback method(s) at which the `validation_data` and therewith its size is available to the `Callback` is the ones that are called at each epoch. Because it's most likely not desirable to create the `tf.train.Saver...` and the `ProjectorConfig` over and over again, I don't see a way to simplify the interface.

dschwertfeger commented:
I finally got around to looking into this again. I got closer to solving the issue of processing in batches. As you described in your comment from Oct 13, 2017, @chleibig, feeding the `batch_id` and assigning slices `tf.assign(embedding[batch_id:batch_id + self.batch_size], embedding_input)` works fine **if the `embedding_input` is also reshaped accordingly**. 

```python
embedding_input = tf.reshape(embedding_input, (self.batch_size, embedding_size))
```

That presents another another challenge because the last batch might (and very often is) smaller than the `batch_size`, in which case the reshape operation complains.

Any ideas or pointers how to deal with this issue?

chleibig commented:
In [my comment](https://github.com/keras-team/keras/pull/7766#discussion_r144500220), I had this scenario of course in mind, for which reason I had written it with `step` instead of `self.batch_size`. Step had actually referred to your code: `step = min(self.batch_size, n_samples - i)` I guess. Long story short, assign as many samples as you have in your current batch.

dschwertfeger commented:
I completely overlooked that. I had renamed `step` to `offset` locally a while ago and already wondered where your magical `step` should come from. It makes sense and works now, though. I'll clean this up and push tomorrow.

chleibig commented on keras/callbacks.py:
```diff
@@ -779,32 +785,46 @@ def is_indexed_slices(grad):
             if not embeddings_layer_names:
                 embeddings_layer_names = [layer.name for layer in self.model.layers
                                           if type(layer).__name__ == 'Embedding']
+            self.embeddings = []
```
just a matter of style and not so important if things work, but lot's of variables from here on have names involving "embedding" - do we really need that many? And if so, are more expressive names conceivable?

dschwertfeger commented:
Thanks for your review, @chleibig.

I'll certainly look into streamlining the use of variables and updating documentation. I feel the use-case of multi-input models deserves more discussion, though.

As [I've mentioned before](https://github.com/keras-team/keras/pull/7766#discussion_r143716897), I'm not convinced that embedding multiple inputs in arbitrary layers yields useful results. 

Maybe I don't understand this well enough but my hunch is that for models with multimodal inputs, say images and their labels, you need to actually *construct* a joint embedding space rather than simply send data through the network and take the output of arbitrary layers as the "embedding" (see [1](https://arxiv.org/pdf/1604.04573.pdf), [2](https://arxiv.org/pdf/1511.06078.pdf) for examples).

What exactly would this embedding then represent?

Is it sensible to make something work, technically, which doesn't have practical relevance (or meaning)?

This PR started as a bug fix for a broken feature. I'd really like to wrap this up and I'm wondering if we should tackle multi-input as a separate issue or feature.

What's your view on that?

chleibig commented:
Thanks for the references @dschwertfeger and I also would like to close this issue, but in the current state, the callback will fail if used together with a model with multiple inputs. Please note that other parts of the `TensorBoard` callback do take multiple inputs into account!

Thinking about the practical relevance of embeddings is certainly interesting, but is a bit out of scope for this PR. Anyway, I don't see any reason why "obtaining hidden layer activations from a model and exporting them for performing low-dim embeddings in Tensorboard" (this is what this PR introduces) would be of less practical relevance if there are multiple inputs to the model. The embeddings are determined, by the problem domain, objective, ...

tboquet commented on keras/callbacks.py:
```diff
@@ -779,32 +785,46 @@ def is_indexed_slices(grad):
             self.writer = tf.summary.FileWriter(self.log_dir)
 
-        if self.embeddings_freq:
+        if self.embeddings_freq and self.embeddings_data is not None:
```
Let's say we only want to visualize the weights, should we always pass data?

dschwertfeger replied:
This PR's intention is to fix the callback from only visualizing the weights to actually embedding the data using the weights. Visualizing the weights won't be possible with these changes.

tboquet replied:
Ah, I guess we should be able to do both. Removing this feature could impact other users. @chleibig @fchollet any thoughts on that?

chleibig replied:
I wouldn't miss the original functionality, but that's just my personal opinion...revising the conversation of the original PR that had introduced the weight embedding functionality might reveal whether this is needed or indeed was unintentionally introduced.

tboquet commented on keras/callbacks.py:
```diff
@@ -779,32 +785,46 @@ def is_indexed_slices(grad):
 
-            self.saver = tf.train.Saver(list(embeddings.values()))
+                    embedding_input = self.model.get_layer(layer.name).output
+                    embedding_size = reduce(mul, embedding_input.shape[1:])
```
Can `np.prod` be used in this case?

dschwertfeger replied:
Yes, would you prefer that?

tboquet replied:
it's less pythonic but more in line with the style of how things are computed at some other places in the code base.

dschwertfeger replied:
Okay. I thought of using standard lib when possible but I also get that it's good practice to be in line with the rest of the code base. `np.prod` it is, then.

tboquet commented on keras/callbacks.py:
```diff
@@ -848,11 +871,35 @@ def on_epoch_end(self, epoch, logs=None):
+                    batch_val = embeddings_data[i:i + step]
+
+                    if type(self.model.input) == list:
+                        feed_dict = {model_input: batch_val
```
Agreed, this should be taken care of, we should be able to pass a list of numpy arrays.

tboquet commented on keras/callbacks.py:
```diff
@@ -779,32 +785,46 @@ def is_indexed_slices(grad):
+                    embedding_input = tf.reshape(embedding_input,
+                                                 (step, int(embedding_size)))
+                    shape = (self.embeddings_data.shape[0], int(embedding_size))
+                    embedding = tf.Variable(tf.zeros(shape),
```
Is it possible to use Keras' backend here?

dschwertfeger replied:
It's probably possible but unnecessary because this is all inside the TensorBoard callback which only works with TF backend.

tboquet replied:
I agree that it will only work with tensorflow but in terms of style it will improve the quality of your PR. It limits the use of tensorflow outside of the backend.

dschwertfeger replied:
Not sure what you mean by

> It limits the use of tensorflow outside of the backend.

Could you please rephrase or elaborate?

Why would you consider it better style to use the generic Keras backend although this code can never be used with any other backend then the TF backend?

Can we maybe have a third opinion on that, @chleibig ?

tboquet replied:
The idea was: if we can use Keras' TF backend we should use it, if we cannot we fall back to native TF. This way we have less backend specific code using native TF outside of the Keras bakend (even if we can only use this callback only with the TF backend and for some parts we use native TF). It's more about style than usability. I would be the first time a `tf.Variable` is used outside of the backend.
Since we have to use tensorflow to write to the logs and we don't use the extra attributes given to the tensors it's not that important. But I would wait for someone else to comment on that.

chleibig replied:
I totally agree with @tboquet that it would be better style but I would only do it if all `tf` statements can be avoided in this TF specific callback. Otherwise we still have a mixture of backend and native statements.

dschwertfeger replied:
We can't get rid of all the `tf`-statements because of the `FileWriter` and `Saver` and other `tf`-specific stuff which is not available on Keras' backend.

dschwertfeger commented on keras/callbacks.py:
```diff
@@ -658,14 +660,16 @@ class TensorBoard(Callback):
             [details](https://www.tensorflow.org/how_tos/embedding_viz/#metadata_optional)
             about metadata files format. In case if the same metadata file is
             used for all embedding layers, string can be passed.
+        embeddings_data: embedding data as Numpy array.
```
I would like to use `_standardize_input_data` but it lives in `keras.engine.training` which already imports from `callbacks`. So, I'm running into a circular import problem. 

I couldn't find any other useful function for input validation either.

What do you suggest to do? Should we move `_standardize_input_data` somewhere else? If so, where to?

At the moment it's only used `keras.engine.training`

dschwertfeger replied:
Thoughts, @chleibig ?

chleibig replied:
If you indeed want to use this function, this sounds reasonable. In order to keep similar things together, I would suggest to move all the `*standardize*` functions to an util module, probably as part of the `engine` package. Any objections @fchollet?

dschwertfeger commented:
I have addressed your feedback and believe we can move this forward now. @chleibig, would you mind reviewing the latest changes and see if this can be approved for merging?

ahundt commented:
These changes look much improved from my perspective. I'd only suggest adding a small `tensorboard_embeddings_mnist.py` example or something so users have a starting point for this functionality. Would that be easy enough to do?

dschwertfeger commented:
Sure, I'll take a shot at it.

dschwertfeger commented:
Quick update:

- I added an example, 
- I solved the merge-conflicts from latest changes in master
- The test are green

What's missing now is one approving review. Could anyone please review this again?

chleibig commented:
Sounds great, I'll have a look this weekend.

dschwertfeger commented on keras/callbacks.py:
```diff
@@ -446,8 +447,8 @@ def on_epoch_end(self, epoch, logs=None):
-                            print('\nEpoch %05d: %s did not improve from %0.5f' %
-                                  (epoch + 1, self.monitor, self.best))
+                            print('\nEpoch %05d: %s did not improve' %
+                                  (epoch + 1, self.monitor))
```
Oops. No, this is a mistake that happened when I merged the latest master. Good catch.

dschwertfeger commented on keras/callbacks.py:
```diff
@@ -856,11 +884,34 @@ def on_epoch_end(self, epoch, logs=None):
+                n_samples = embeddings_data[0].shape[0]
+
+                i = 0
+                while i < n_samples:
```
Left comment above. Hope that's clear enough.

ahundt commented:
@dschwertfeger keras master is building again, can you merge? hopefully re-run of travis will then work correctly

dschwertfeger commented:
The tests are green now and we have two approving reviews. Merging is still blocked though because of one pending review.

@tboquet, could you please take a look again?

dschwertfeger commented:
Thanks!

This PR has now 3 approving reviews but merging is still blocked.

> Review required: At least 1 approving review is required by reviewers with write access

Looping in @fchollet again to make the final call to merge. 

@fchollet, as a reminder, in https://github.com/keras-team/keras/pull/7766#issuecomment-363542356 you said you'd wait for @ahundt's and @chleibig's approval to merge this. 

They have both approved the changes and the PR is ready to be merged.

I'd like to thank all of you for guiding me through this work and helping when I was stuck!

ahundt commented:
@fchollet I think this is good to go! Thanks for all the effort @dschwertfeger

drcege commented:
If I understand correctly, the callback requires an input tensor (data to be embedded) and the weights of a embedding layer (to embed the data). So is there a way to visualize some vectors immediately. 

For example, I use the pre-trained word embedding vectors as an embedding space (the first input to the model) and embed some words into the same space (the second input to the model).  The pre-trained word embeddings are numpy arrays and only be fed to compute the loss. So it is not the output of a layer but just placeholder.  I can visualize the embedded words, but how to visualize them together?

phisad commented:
Are there any consideration to make `embeddings_data` work with Keras sequences or generators?

Looking into the code I only see support for `dict`, `list` and `DataFrame`.

I'd really like to see Keras sequences to work here because its a Keras provided way to handle data (which cannot be totally loaded in memory).
